analyzers/src/SonarAnalyzer.Common/Rules/UnusedStringBuilderBase.cs

```diff
@@ -0,0 +1,75 @@
+{
+    private const string DiagnosticId = "S3063";
+    protected override string MessageFormat => """Remove this "StringBuilder"; ".ToString()" is never called.""";
+
+    internal readonly string[] StringBuilderAccessInvocations = { "ToString", "CopyTo", "GetChunks" };
+    internal readonly string[] StringBuilderAccessExpressions = { "Length", "Capacity", "MaxCapacity" };
+
+    protected abstract ILocalSymbol GetSymbol(TVariableDeclarator declaration, SemanticModel semanticModel);
+    protected abstract string GetName(SyntaxNode declaration);
+    protected abstract SyntaxNode GetScope(TVariableDeclarator declarator);
+    protected abstract bool NeedsToTrack(TVariableDeclarator declaration, SemanticModel semanticModel);
+    protected abstract bool IsStringBuilderRead(string name, ILocalSymbol symbol, SyntaxNode node, SemanticModel model);
+    protected abstract bool DescendIntoChildren(SyntaxNode node);
+
+    protected UnusedStringBuilderBase() : base(DiagnosticId) { }
+
+    protected sealed override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            var variableDeclaration = (TVariableDeclarator)c.Node;
+            if (!NeedsToTrack(variableDeclaration, c.SemanticModel)
+                || GetScope(variableDeclaration).DescendantNodes(DescendIntoChildren).Any(node =>
+                    IsStringBuilderRead(GetName(variableDeclaration), GetSymbol(variableDeclaration, c.SemanticModel), node, c.SemanticModel)))
+            {
+                return;
+            }
+            c.ReportIssue(Diagnostic.Create(Rule, variableDeclaration.GetLocation()));
+        }, Language.SyntaxKind.VariableDeclarator);
+
+    internal bool IsSameReference(SyntaxNode expression, string name, ILocalSymbol symbol, SemanticModel semanticModel)
+    {
+        var references = GetLocalReferences(expression, semanticModel);
+        if (!references.Any() && expression.Ancestors().OfType<TConditionalExpression>().Any())
+        {
+            references = GetLocalReferences(expression.Ancestors().OfType<TConditionalExpression>().First(), semanticModel);
+        }
+        return references.Any(x => IsSameVariable(x, name, symbol, semanticModel));
+    }
+
+    internal bool IsSameVariable(SyntaxNode identifier, string name, ILocalSymbol symbol, SemanticModel semanticModel) =>
+        GetName(identifier).Equals(name, Language.NameComparison) && symbol.Equals(semanticModel.GetSymbolInfo(identifier).Symbol);
+
+    internal static IEnumerable<TIdentifierName> GetLocalReferences(SyntaxNode node, SemanticModel semanticModel) =>
+        node.DescendantNodesAndSelf().OfType<TIdentifierName>().Where(x => IsLocalReference(x, semanticModel));
+
+    internal static bool IsLocalReference(SyntaxNode identifier, SemanticModel semanticModel) =>
+        semanticModel.GetOperation(identifier) is { Kind: OperationKindEx.LocalReference };
```

Comment:
`semanticModel.GetOperation` is even more expensive than other semanticModel calls. We, therefore, try to avoid calling it if possible.
We can live with false positives in expression contexts (almost everything is an expression and there are just too many cases to cover all).
It is enough to just retrieve all the IdentifierNames, compare the name syntactically, and make sure we have the same symbol to avoid mismatches caused by shadowing.

Start with the argument syntax or return syntax and so on and inspect all the identifiers found in their expressions (`expression.DescendantNodesAndSelf.OfType<IdentifierNameSyntax>()`) (this is close to what you do right now). There shouldn't be any need to special case any expression kinds like TConditionalExpression.

Reply:
This was changed during the review to fix the AD0001 on ITs.
Some project was throwing a an `System.InvalidOperationException` with message `Feature 'IOperation' is disabled.`.
Older projects do not have the [IOperation](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.ioperation?view=roslyn-dotnet-4.3.0) feature (or not enabled by default). Anyway as you said that wasn't as performant as it can be, that part is now removed.